sentry-core/src/main/java/io/sentry/core/transport/HttpTransport.java

```diff
@@ -158,38 +160,19 @@ public HttpTransport(
 
     int responseCode = -1;
 
-    try (final OutputStream outputStream = connection.getOutputStream();
-        final Writer writer = new BufferedWriter(new OutputStreamWriter(outputStream, UTF_8))) {
+    try (final ByteArrayOutputStream byteOut = new ByteArrayOutputStream();
+        final OutputStream outputStream = connection.getOutputStream();
+        final GZIPOutputStream gzip = new GZIPOutputStream(byteOut);
+        final Writer writer = new BufferedWriter(new OutputStreamWriter(gzip, UTF_8))) {
```

Comment:
I'm not familiar with the stack used in the Java SDK, so perhaps the best I can contribute is to ask whether it is possible / you considered plugging in streams directly instead of using a new byte array buffer for every event?

I suspect it would create a lot of garbage. Maybe not too relevant if sending occasional errors, but can be relevant for code running more often like transactions and maybe sessions (not very familiar with how sessions work though).

In Go, it would be a good use case for [`sync.Pool`](https://golang.org/pkg/sync/#Pool), which is a way to amortize allocation overhead. There must be something similar in Java.

So there is really two independent considerations here:

1. Streaming without in-memory buffering
2. Object reuse / pools to amortize allocation overhead

Reply:
[in .NET we have an option to use buffered or not](https://github.com/getsentry/sentry-dotnet/blob/04d756b024a2ee00bb5d2748630ba704349b4f88/src/Sentry/Internal/Http/DefaultSentryHttpClientFactory.cs#L76-L87). This came in after some environments that couldn't use the non-buffered approach IIRC.

No byte array is used directly here so that we can swap out with a pooled version, or am I missing something?

`ByteArrayOutputStream` can be using object pooling under the hood though.

Reply:
I'll have a look if its possible to make a bit more optimal, unfortunately, we are very limited on JDK 6 usage on many things due to the Android runtime, otherwise, we'd need to have transport for Android and one for sentry-java.

Object pooling is not a thing for Java and there are a few things out there but we can't really use it. eg https://commons.apache.org/proper/commons-pool/ or Streams, Slices, etc..

gues I can also use `BufferedOutputStream` instead of `ByteArrayOutputStream` directly.
Or even changing the serialization layer to write directly to this stream, the later would require a bigger change.

let's what what I can find, thanks for the review btw.

Reply:
ok so no `ByteArrayOutputStream` anymore, it writes directly in the stream of the connection.
`GZIPOutputStream` already has its own buffering and I'll keep its default values.

The session works this way: a session starts when the user opens the App and ends when the user goes to the background and remains idle for a period, so no worries here, it's by far not as much as transactions.

Maybe in a later stage, we might need to refactor this a little bit to be more optimal, but we'll be ended up with different transports for Mobile and Backend, as we can't be cheap and effective at the same time (eg right now we open/close connections for each request), this is ~good for mobile, but might not be good for the backend.

This is a lot better than before, thanks for the feedback, if no more concerns, I'd like to resolve this issue.